File: sdk/AuthSys-Device-Example/rust/AuthSys-Device.rs
use std::collections::HashMap;
use std::time::Duration;
use reqwest::blocking::Client;

pub struct Device {
    app_secret: String,
    base_url: String,
    client: Client,
    pub last_error: String,
    pub last_response: String,
}

impl Device {
    pub fn new(app_secret: &str, base_url: &str) -> Self {
        let url = if base_url.is_empty() {
            "https://authsys-main-production.up.railway.app/device".to_string()
        } else {
            base_url.trim_end_matches('/').to_string()
        };
        let client = Client::builder()
            .timeout(Duration::from_secs(15))
            .danger_accept_invalid_certs(true)
            .build()
            .unwrap();
        Device {
            app_secret: app_secret.to_string(),
            base_url: url,
            client,
            last_error: String::new(),
            last_response: String::new(),
        }
    }

    fn get_hwid() -> String {
        let hostname = hostname::get()
            .map(|h| h.to_string_lossy().to_string())
            .unwrap_or_default();
        let hash = format!("{:X}", md5::compute(hostname.as_bytes()));
        hash
    }

    fn request(&mut self, endpoint: &str, payload: &HashMap<&str, &str>) -> Option<serde_json::Value> {
        let url = format!("{}/{}", self.base_url, endpoint);
        match self.client.post(&url).json(payload).send() {
            Ok(resp) => {
                self.last_response = resp.text().unwrap_or_default();
                serde_json::from_str(&self.last_response).ok()
            }
            Err(e) => {
                self.last_error = e.to_string();
                None
            }
        }
    }

    pub fn check(&mut self) -> bool {
        self.last_error.clear();
        let mut payload = HashMap::new();
        payload.insert("app_secret", &self.app_secret[..]);
        payload.insert("hwid", &Self::get_hwid()[..]);

        match self.request("check", &payload) {
            Some(data) => {
                if data["active"].as_bool().unwrap_or(false) {
                    true
                } else {
                    self.last_error = data["message"]
                        .as_str()
                        .unwrap_or("Device deactivated by admin")
                        .to_string();
                    false
                }
            }
            None => false,
        }
    }

    pub fn register(&mut self, device_name: &str) -> bool {
        self.last_error.clear();
        let mut payload = HashMap::new();
        payload.insert("app_secret", &self.app_secret[..]);
        payload.insert("hwid", &Self::get_hwid()[..]);
        if !device_name.is_empty() {
            payload.insert("device_name", device_name);
        }

        match self.request("register", &payload) {
            Some(data) => data["active"].as_bool().unwrap_or(false),
            None => false,
        }
    }
}

File: sdk/AuthSys-Device-Example/java/AuthSys-Device.java
import java.io.*;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.util.Scanner;

public class AuthSysDevice {

    private final String appSecret;
    private final String baseUrl;
    public String lastError = "";
    public String lastResponse = "";

    public AuthSysDevice(String appSecret) {
        this(appSecret, "https://authsys-main-production.up.railway.app/device");
    }

    public AuthSysDevice(String appSecret, String baseUrl) {
        this.appSecret = appSecret;
        this.baseUrl = baseUrl.replaceAll("/+$", "");
    }

    private static String getHWID() {
        try {
            StringBuilder raw = new StringBuilder();
            raw.append(java.net.InetAddress.getLocalHost().getHostName());
            raw.append(System.getProperty("os.name"));
            raw.append(System.getProperty("os.version"));
            raw.append(System.getProperty("os.arch"));
            try {
                java.net.NetworkInterface ni = java.net.NetworkInterface.getByInetAddress(
                    java.net.InetAddress.getLocalHost());
                if (ni != null) {
                    byte[] mac = ni.getHardwareAddress();
                    if (mac != null) {
                        for (byte b : mac) raw.append(String.format("%02X", b));
                    }
                }
            } catch (Exception ignored) {}
            MessageDigest md = MessageDigest.getInstance("MD5");
            byte[] hash = md.digest(raw.toString().getBytes());
            StringBuilder sb = new StringBuilder();
            for (byte b : hash) sb.append(String.format("%02X", b));
            return sb.toString();
        } catch (Exception e) {
            return "unknown";
        }
    }

    private String postRequest(String endpoint, String json) throws IOException {
        URL url = new URL(baseUrl + "/" + endpoint);
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        conn.setDoOutput(true);
        conn.setConnectTimeout(15000);
        conn.setReadTimeout(15000);

        try (OutputStream os = conn.getOutputStream()) {
            byte[] input = json.getBytes(StandardCharsets.UTF_8);
            os.write(input, 0, input.length);
        }

        int status = conn.getResponseCode();
        Scanner scanner;
        if (status >= 200 && status < 300) {
            scanner = new Scanner(conn.getInputStream());
        } else {
            scanner = new Scanner(conn.getErrorStream());
        }
        String body = scanner.useDelimiter("\\A").hasNext() ? scanner.next() : "";
        scanner.close();
        conn.disconnect();
        return body;
    }

    private String jsonGet(String json, String key) {
        String search = "\"" + key + "\":";
        int idx = json.indexOf(search);
        if (idx == -1) return "";
        int start = idx + search.length();
        if (json.charAt(start) == '"') {
            start++;
            int end = json.indexOf('"', start);
            return end == -1 ? "" : json.substring(start, end);
        }
        int end = json.indexOf(",", start);
        if (end == -1) end = json.indexOf("}", start);
        return end == -1 ? json.substring(start) : json.substring(start, end).trim();
    }

    public boolean check() {
        lastError = "";
        try {
            String json = "{\"group_secret\":\"" + appSecret + "\",\"hwid\":\"" + getHWID() + "\"}";
            lastResponse = postRequest("check", json);
            String active = jsonGet(lastResponse, "active");
            if ("true".equals(active)) return true;
            String msg = jsonGet(lastResponse, "message");
            lastError = msg.isEmpty() ? "Device deactivated by admin" : msg;
            return false;
        } catch (Exception e) {
            lastError = e.getMessage();
            return false;
        }
    }

    public boolean register(String deviceName) {
        lastError = "";
        try {
            String json = "{\"group_secret\":\"" + appSecret + "\",\"hwid\":\"" + getHWID() + "\"";
            if (deviceName != null && !deviceName.isEmpty()) {
                json += ",\"device_name\":\"" + deviceName + "\"";
            }
            json += "}";
            lastResponse = postRequest("register", json);
            String active = jsonGet(lastResponse, "active");
            return "true".equals(active);
        } catch (Exception e) {
            lastError = e.getMessage();
            return false;
        }
    }
}
